### ISSUE TITLE
Improve address detection for remote execution

### ISSUE THREAD

**Issue by yadudoc** [Parsl/parsl#869]
Currently we provide a few different mechanisms for users to detect an address/interface to which users can connect to (#466). However these mechanisms only gets us halfway as far as usability is concerned. Often times, the user has to do a fair bit of trial and error to find the right address when multiple interface are available. This can get complicated when there's only a specific port-range that is open. Ideally we have a mechanism to probe available addresses and use one that works.

To start off, we can try adding this to `HighThroughputExecutor` by adding support to the manager code (process_worker_pool.py) so that the manager would try a list of addresses if the default fails.

**Comment by timarmstrong** [comment]
@yadudoc do you have any pointers for how I might test this? Are there parsl tests that simulate this kind of scenario?

One thought is that you could emulate this on a single node with docker bridge networks, but I wouldn't want to drag in docker as a test dependency just for this. I recently have been doing some work in Impala running tests like this so it's fresh in my mind. I had HDFS/HBase/Hive Metastore/etc running on the host and listening only on the docker bridge network and other processes running in the docker bridge network and connecting back. It's a reasonable approximation, I think, of the scenarios you're talking about.

**Comment by benclifford** [comment]
Maybe try passing in a list of three addresses, the first and last of which you don't expect to work. Eg `[8.8.8.8, 127.0.0.1, 8.8.8.7]`.

Docker might be appropriate in a larger project to test fault handling in parsl - I've certainly used it that was on other projects. Maybe that should be a separate issue to track.

**Comment by yadudoc** [comment]
We have a bunch of remote configurations (`parsl/parsl/tests/configs`) that are tested manually but those might not be much use without access to those remote systems. The docker route sounds reasonable.  As long as we have multiple interfaces out of which at least one doesn't work we can test this. We want to set a bad address manually, and have the executor send alternative addresses to the manager to fallback on.

```python
config = Config(
    executors=[
        HighThroughputExecutor(
            address=#BAD_ADDRESS_HERE, <--------------------- Add this manually
            label="htex_local",
            worker_debug=True,                                                       
            cores_per_worker=1,
            provider=LocalProvider(
                channel=LocalChannel(),
                init_blocks=1,  
                # ^ This will start 1 manager, with as many workers as there are cores
                max_blocks=1,                                                                                                       
                launcher=SingleNodeLauncher(),
            ),
        )
    ],
    strategy=None,
)
```

**Comment by benclifford** [comment]
@yadudoc are you imagining that HT gets initialised with one address but that it also acquires a load of other addresses outside of the configuration? I was imagining that you would pass in a whole list of addresses, aquired by some new helper function (eg one that combines the output of all of the existing helpers). That feels more debuggable to me.

**Comment by yadudoc** [comment]
Right. If the user specifies the `address` we use that, and if that fails falls back to a list of parsl-generated addresses from helpers. I'd think you want this to be automatic, but also allow a full manual mode where if the user specified the address we'd use that first. I wanted to avoid a situation where the preferred address is specified via the order in `addresses` field.

**Comment by benclifford** [comment]
Software that decides the address you gave it is bad and goes off to do its own thing secretly can be a sysadmin nightmare.

I'd much rather this default to something like generating the list automatically, but if you override that with an explicit address list, that specified list is the address list that is used and nothing else.

If the automatic list generation is going to catch most cases, then this should be ok: people will almost never need to specify an address list explicitly.

**Comment by yadudoc** [comment]
Just to be clear, don't these defaults cover your cases : 
`Executor(address=None, addresses=get_all_addresses())`
1. If user specifies nothing, we try everything we could find. <- Full auto
2. If the user specified a working address, we try that first, and falls back to the default list (if there's one). <- Manual with fallback on auto
3. If the user wants an alternate list/ordering, they can change the defaults. <- Full manual.

We are not overriding the user, but rather trying alternatives, that too only when the user supplied address is bad.

**Comment by benclifford** [comment]
Yeah if those two parameters are the only source of addresses those defaults are fine. I'm even fine with None being supplied for both triggering an internal call to get all addresses.
I'm less fine with there being two address parameters.

**Comment by annawoodard** [comment]
I favor simplifying to just one address parameter:
```
Executor(addresses=None) # case 1: we set addresses to get_all_addresses()

Executor(addresses=['foo']) # case 2a: user has specified one or more addresses, we honor their list

Executor(addresses=['foo', 'bar'] + get_all_addresses()) # case 2b: if needed, user can configure manual with fallback on auto, but at least it is explicit
```
It sacrifices a small amount of automatic-ness but makes things simpler and reduces mental load on someone trying to debug failures, or understand how to properly configure  the executor.

**Comment by benclifford** [comment]
this should be implemented by #1422
